File: src/models/summary.py
"""Summary data models."""

from datetime import UTC, datetime

from pydantic import BaseModel, Field
from sqlalchemy import JSON, Column, DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.orm import relationship

from src.models.base import Base


class Summary(Base):
    """Content summary database model.

    Summaries are linked to Content records via content_id.
    """

    __tablename__ = "summaries"

    id = Column(Integer, primary_key=True, autoincrement=True)
    # Content FK - links summary to source content
    content_id = Column(
        Integer, ForeignKey("contents.id", ondelete="SET NULL"), nullable=True, index=True
    )

    # Summary content
    executive_summary = Column(Text, nullable=False)
    key_themes = Column(JSON, nullable=False)  # List[str]
    strategic_insights = Column(JSON, nullable=False)  # List[str]
    technical_details = Column(JSON, nullable=False)  # List[str]
    actionable_items = Column(JSON, nullable=False)  # List[str]
    notable_quotes = Column(JSON, nullable=False)  # List[str]
    relevant_links = Column(
        JSON, nullable=False, default=list
    )  # List[Dict[str, str]] - {"title": "...", "url": "..."}

    # Relevance scoring
    relevance_scores = Column(JSON, nullable=False)  # Dict[str, float]

    # Unified content model fields (Phase 4)
    markdown_content = Column(Text, nullable=True)  # Full markdown representation
    theme_tags = Column(JSON, nullable=True)  # List[str] - Extracted theme tags

    # Metadata
    agent_framework = Column(String(100), nullable=False)  # claude, openai, google, microsoft
    model_used = Column(String(100), nullable=False)  # General model ID (e.g., "claude-sonnet-4-5")
    model_version = Column(String(20), nullable=True)  # Version (e.g., "20250929")
    created_at = Column(DateTime, nullable=False, default=lambda: datetime.now(UTC))
    token_usage = Column(Integer, nullable=True)
    processing_time_seconds = Column(Float, nullable=True)

    # Relationships
    content = relationship("Content", back_populates="summaries")


# Backwards compatibility alias (deprecated)
NewsletterSummary = Summary


class SummaryData(BaseModel):
    """Pydantic model for summary data transfer."""

    content_id: int | None = None  # FK to contents table
    executive_summary: str
    key_themes: list[str] = Field(default_factory=list)
    strategic_insights: list[str] = Field(default_factory=list)
    technical_details: list[str] = Field(default_factory=list)
    actionable_items: list[str] = Field(default_factory=list)
    notable_quotes: list[str] = Field(default_factory=list)
    relevant_links: list[dict[str, str]] = Field(
        default_factory=list
    )  # [{"title": "...", "url": "..."}]
    relevance_scores: dict[str, float] = Field(default_factory=dict)

    # Unified content model fields
    markdown_content: str | None = None
    theme_tags: list[str] | None = None

    agent_framework: str
    model_used: str  # General model ID (e.g., "claude-sonnet-4-5")
    model_version: str | None = None  # Version (e.g., "20250929")
    token_usage: int | None = None
    processing_time_seconds: float | None = None

File: web/src/types/summary.ts
/**
 * Summary Types
 *
 * TypeScript interfaces for content summary entities.
 * Summaries are linked to Content records (unified model).
 *
 * @see Backend model: src/models/summary.py
 */

/**
 * Content Summary entity
 *
 * Contains structured information extracted from content by the LLM.
 * Each content has at most one summary (1:1 relationship).
 */
export interface NewsletterSummary {
  /** Unique identifier */
  id: number

  /** ID of the content this summary belongs to */
  content_id: number

  /**
   * Executive summary - 2-3 sentence high-level overview
   * Suitable for leadership or quick scanning
   */
  executive_summary: string

  /**
   * Key themes identified in the content
   * Array of theme names/topics covered
   */
  key_themes: string[]

  /**
   * Strategic insights for business leaders
   * Higher-level implications and recommendations
   */
  strategic_insights: string[]

  /**
   * Technical details for practitioners
   * Implementation notes, code references, technical depth
   */
  technical_details: string[]

  /**
   * Actionable items extracted from the content
   * Things readers could implement or investigate
   */
  actionable_items: string[]

  /**
   * Notable quotes from the content
   * Interesting or impactful statements
   */
  notable_quotes: string[]

  /**
   * Relevant links for further reading
   * URLs mentioned in the content worth highlighting
   */
  relevant_links: Array<Record<string, unknown>>

  /**
   * Relevance scores for different audience segments
   */
  relevance_scores: {
    cto_leadership: number
    technical_teams: number
    individual_developers: number
  }

  /**
   * Agent framework used for summarization
   * e.g., "claude-sdk", "langchain", "autogen"
   */
  agent_framework: string

  /**
   * LLM model identifier used
   * e.g., "claude-haiku-4-5", "gpt-4o"
   */
  model_used: string

  /**
   * Specific model version (if available)
   */
  model_version: string | null

  /** When the summary was created */
  created_at: string // ISO 8601 date string

  /** Total tokens used in generation */
  token_usage: number | null

  /** Time taken to generate in seconds */
  processing_time_seconds: number | null
}

/**
 * Summary list item (lightweight view)
 *
 * For displaying in lists without full content.
 */
export interface SummaryListItem {
  id: number
  /** Content ID this summary belongs to */
  content_id: number
  /** Content title for display */
  title: string
  /** Content publication name */
  publication: string | null
  /** First 200 chars of executive summary */
  executive_summary_preview: string
  key_themes: string[]
  model_used: string
  created_at: string
  processing_time_seconds: number | null
}

/**
 * Progress event for summarization task (SSE)
 */
export interface SummarizationProgress {
  /** Task identifier */
  task_id: string
  /** Current step description */
  step?: string
  /** Progress percentage (0-100) */
  progress: number
  /** Currently processing content ID */
  current_content_id?: number
  /** Count of completed summaries */
  completed_count?: number
  /** Total count to process */
  total_count?: number
  /** Count of processed */
  processed?: number
  /** Total */
  total?: number
  /** Status: processing, completed, or error */
  status: "queued" | "processing" | "completed" | "error"
  /** Message from the server */
  message?: string
}

/**
 * Filters for summary list queries
 */
export interface SummaryFilters {
  /** Filter by content ID */
  content_id?: number
  /** Filter by model used */
  model_used?: string
  /** Filter after this date */
  start_date?: string
  /** Filter before this date */
  end_date?: string
  /** Pagination limit */
  limit?: number
  /** Pagination offset */
  offset?: number
}

/**
 * Navigation response for prev/next within a filtered list
 */
export interface SummaryNavigation {
  prev_id: number | null
  next_id: number | null
  prev_content_id: number | null
  next_content_id: number | null
  position: number
  total: number
}
